src/mc-tokens-private.h

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2022-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MONGOCRYPT_TOKENS_PRIVATE_H
+#define MONGOCRYPT_TOKENS_PRIVATE_H
+
+#include "mongocrypt-buffer-private.h"
+#include "mongocrypt-crypto-private.h"
+
+/*
+ * ======================= Begin: FLE 2 Token Reference =======================
+ *
+ * v is a BSON value. It is the bytes after "e_name" in "element" in
+ * https://bsonspec.org/spec.html.
+ * u is a "contention factor" counter. It is a uint64_t.
+ * HMAC is the HMAC-SHA-256 function.
+ * Integers are represented as uint64_t in little-endian.
+ *
+ * CollectionsLevel1Token = HMAC(RootKey, 1)
+ * ServerDataEncryptionLevel1Token = HMAC(RootKey, 3)
+ *
+ * EDCToken = HMAC(CollectionsLevel1Token, 1)
+ * ESCToken = HMAC(CollectionsLevel1Token, 2)
+ * ECCToken = HMAC(CollectionsLevel1Token, 3)
+ * ECOCToken = HMAC(CollectionsLevel1Token, 4)
+ *
+ * EDCDerivedFromDataToken = HMAC(EDCToken, v)
+ * ESCDerivedFromDataToken = HMAC(ESCToken, v)
+ * ECCDerivedFromDataToken = HMAC(ECCToken, v)
+ *
+ * EDCDerivedFromDataTokenAndCounter = HMAC(EDCDerivedFromDataToken, u)
+ * ESCDerivedFromDataTokenAndCounter = HMAC(ESCDerivedFromDataToken, u)
+ * ECCDerivedFromDataTokenAndCounter = HMAC(ECCDerivedFromDataToken, u)
+ * ======================== End: FLE 2 Token Reference ========================
+ */
+
+
+typedef struct _mc_CollectionsLevel1Token_t mc_CollectionsLevel1Token_t;
```

Comment:
A lot of this looks duplicated. I played around and created two preprocessor macros that emit equivalent code, as far as I can tell. I've attached a patch with the difference.
[macros.txt](https://github.com/mongodb/libmongocrypt/files/8276616/macros.txt)